packages/core/contracts-ovm/insured-bridge/implementation/BridgePool.sol

```diff
@@ -125,108 +123,154 @@ contract BridgePool is Testable {
-
-        // TODO: There is more data we'd like to emit, such as depositId, but we are limited by how many variables
-        // we can fit into an event. Perhaps we need multiple events? Or think more critically about what information
-        // clients will want.
-        emit DepositRelayed(
-            l2Sender,
-            depositTimestamp,
-            recipient,
+        // Note: We need to set a new price request timestamp instead of default setting to equal to the
+        // `depositTimestamp`, which is dependent on the L2 VM on which the DepositContract is deployed. Imagine if
+        // the timestamps on the L2 have an offset that are always "in the future" relative to L1 blocks, then the
+        // OptimisticOracle would always reject requests.
+        _requestOraclePriceRelay(
             l1Token,
-            msg.sender,
             amount,
-            proposerRewardPct,
-            realizedFeePct,
-            bridgeAdmin.depositContract()
+            priceRequestTime,
+            getRelayAncillaryData(depositData, relayData),
+            proposerRewardPct
         );
+        _proposeOraclePriceRelay(l1Token, amount, depositTimestamp, getRelayAncillaryData(depositData, relayData));
+
+        // We use an internal method to emit this event to overcome Solidity's "stack too deep" error.
+        _emitDepositRelayedEvent(depositData, relayAncillaryDataHash, depositHash);
     }
 
-    function speedUpRelay(uint256 depositId) public {}
+    // TODO: Implement fully the following functions:
+    function speedUpRelay(bytes32 _depositHash) public {
+        RelayData storage relay = relays[_depositHash];
+        relay.instantRelayer = msg.sender;
+        // TODO: Pull funds from caller.
+        emit RelaySpedUp(_depositHash, msg.sender);
+    }
+
+    function finalizeRelay(bytes32 _depositHash) public {
+        RelayData storage relay = relays[_depositHash];
+        // TODO: Do stuff using PendingRelay data.
+    }
 
-    function finalizeRelay(uint256 depositId) public {}
+    /**
+     * @notice OptimisticOracle will callback to this function after a pending relay is disputed. This function should
+     * ensure that another slow relayer can fulfill the disputed relay for an L2 deposit.
+     * @dev TODO: Add checks that only the optimistic oracle can call this function.
+     */
+    function priceDisputed(
+        bytes32 identifier,
+        uint256 timestamp,
+        bytes memory ancillaryData,
+        uint256 refund
+    ) public {
+        bytes32 depositHash = ancillaryDataToDepositHash[keccak256(ancillaryData)];
+        RelayData storage relay = relays[depositHash];
 
-    function settleDisputedRelay(uint256 depositId, address slowRelayer) public {}
+        // Mark pending relay as uninitialized but do not delete instant relayer information which should be copied
+        // over to next slow relay.
+        relay.relayState = RelayState.Uninitialized;
```

Comment:
if this has been disputed then `Uninitialized` as a state seems pretty counterintuitive. would making a more explicit `Disputed` State make more sense?

Reply:
also an event would be very useful for this

Reply:
I'll add en event, I was planning to leave implementation for this function for another PR.

> if this has been disputed then Uninitialized as a state seems pretty counterintuitive. would making a more explicit Disputed State make more sense?
The problem with this is that we need to then add a check to relayDeposit from `require(relays[depositHash].relayState == RelayState.Uninitialized, "Pending relay for deposit exists")` to `require(relays[depositHash].relayState == RelayState.Uninitialized || relays[depositHash].relayState == RelayState.Disputed, "Pending relay for deposit exists")`